### ISSUE TITLE
[NEXT-1194] After upgrade to 13.3 request.json() in DELETE route handler fails

### ISSUE THREAD

**Issue by gitdagray** [vercel/next.js#48096]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 18.12.1
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.3.1-canary.1
      eslint-config-next: 13.2.4
      react: 18.2.0
      react-dom: 18.2.0
```

### Link to the code that reproduces this issue

N/A

### To Reproduce

```ts
import { NextResponse } from "next/server"
const DATA_SOURCE_URL = "https://jsonplaceholder.typicode.com/todos"

type Todo = {
    userId: number,
    id: number,
    title: string,
    completed: boolean,
}

export async function DELETE(request: Request) {

    const { id }: Partial<Todo> = await request.json() // error

    if (!id) return NextResponse.json({ "message": "Todo id required" })

    await fetch(`${DATA_SOURCE_URL}/${id}`, {
        method: 'DELETE',
    })

    return NextResponse.json({ "message": `Todo ${id} deleted` })
}
```

### Describe the Bug

I only encounter this error on the DELETE route handler. My POST and PUT route handlers can `await request.json()` with no issues.

I run the same code in 13.2.4 and _do not_ receive the error. 

I am sending valid and very simple JSON: 
```json
{
   "id": 3
}
```

Nevertheless, I receive this error:
```
SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at NextRequest.json (node:internal/deps/undici/undici:6160:23)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async DELETE (webpack-internal:///(sc_server)/./src/app/api/todos/route.ts:63:21)
 ```
    

### Expected Behavior

I expect `await request.json()` to work as it does in the POST and PUT route handlers.

<sub>[NEXT-1194](https://linear.app/vercel/issue/NEXT-1194/after-upgrade-to-133-requestjson-in-delete-route-handler-fails)</sub>

**Comment by fbuyuk** [comment]
+1

**Comment by batuhanbilginn** [comment]
+1

**Comment by niwsa** [comment]
+1 Facing the issue with dev server. Production build works fine.

**Comment by andrewbateman** [comment]
+1

**Comment by nadam131** [comment]
+1

**Comment by zenix-s** [comment]
Similar error I'm using axios, Post works fine but delete dont work and leaves a status code 500

**Comment by arasrezaei** [comment]
Same here... it worked previous versions, now I had to use  url query parameters instead ...

**Comment by simonknittel** [comment]
Still happens with 13.13.1

**Comment by danielmarkow** [comment]
+ 1

**Comment by ralfscheffler** [comment]
Same, same

**Comment by vlsk3104** [comment]
same

**Comment by codingforfoodllc** [comment]
Same here

**Comment by aFendell** [comment]
Same issue

**Comment by daveteu** [comment]
The problem exist also in 13.4.1

**Comment by Carduelis** [comment]
Phew. I'm not alone. 13.4.0

**Comment by Farhanaziz95** [comment]
i dont know if this will also work for delete i got this issue with POST where i wasn't able to get body 
i'm a noob in this kind of this discussions


```
export async function POST(request: Request, response: NextApiResponse) {
    if (request.method !== 'POST') {

    }
    else {

        const a = await request.text()
        const data = JSON.parse(a)
        const { id, email, password, role } = data;
        const modipassword = process.env.SECRATE_KEY + password
        const salt = await bcrypt.genSalt(10)
        const hashedPassword = await bcrypt.hash(modipassword, salt);
        



        const newUser: typeAuthor = {
            id: id,
            name: '',
            email: email,
            password: hashedPassword,
            role: role,
            facebook_link: '',
            insta_link: '',
            twitter_link: '',
            status: 'Active',
            created_at: new Date(),
            updated_at: new Date(),
        };

        await db.insert(Author).values(newUser);



        const custom_data_array = [{ message: 'Successfully Added' }]
        return new Response(JSON.stringify(custom_data_array), { status: 200 })
    }
}
```

well i got work around for my problem like this hope it would be some use to everyone here

**Comment by Frost7994** [comment]
also having this issue with route handler DELETE request, is there any update or simple workaround besides rolling back to an earlier version?

**Comment by alecf** [comment]
I stumbled across this but then found that DELETE generally should not have a request body - the fact that nextjs previously supported it sounds like a bug to me:
https://www.rfc-editor.org/rfc/rfc9110.html#name-delete

Though it sounds frustrating if you were previously depending on this behavior...

**Comment by cjcheshire** [comment]
@alecf nice find, I’ve subsequently changed my request to pass query params and all is good again. Safe to upgrade to latest :)

**Comment by daveteu** [comment]
> @alecf nice find, I’ve subsequently changed my request to pass query params and all is good again. Safe to upgrade to latest :)

How did u manage to do it?

I did the following but i get **Method Not Allowed** error.

```
export async function DELETE(request: Request) {
  const { searchParams } = new URL(request.url);

  const _id = searchParams.get('id') || ""; // for ts
  console.log('_id', _id);
  
  // ...

}
```

**Comment by daveteu** [comment]
@alecf thanks.

I use the following codes for those new to app /api and do not know how to get the query params.

```
DELETE /api/post?_id=12345

export async function DELETE(request: Request) {

  const { searchParams } = new URL(request.url);
  const _id = searchParams.get('_id') || ''; // for ts

  // ...
}
```

**Comment by cjcheshire** [comment]
@daveteu 

```
export async function DELETE(request: Request) {

  const { searchParams } = new URL(request.url);
  const _id = searchParams.get('_id') || ''; // for ts

  // ...
}
```

**Comment by twilio-jyoung** [comment]
Still occurs in Next.js v13.4.2 and is impacting the commerce template: https://github.com/vercel/commerce

**Comment by y-nk** [comment]
Confirming, this also happens in POST requests with the latest canary as of today. Starting to wonder how come there's no testing for this. Feels like regression.

**Comment by Hakeemmidan** [comment]
I'm facing the same issue with a GET request on `next@13.4.3`

**Comment by Frost7994** [comment]
I was having a lot of trouble with this and it turns out the handler gets access to both the request AND context object and the the params comes from the context not request like so:

```
export async function DELETE(req, ctx){
 const {id}=ctx.params;
}
```

**Comment by vutrung96** [comment]
This is still broken for GET requests. Changing to POST fixes the issue

**Comment by iamlucasgomes** [comment]
Ainda com problemas na rota Delete, estou usando next 13.4.4

**Comment by james-william-r** [comment]
Just some clarity on method bodies. It's not common practice for methods other than PUT, POST and PATCH to have a body. 

There's no reason you can't have use a POST request for multiple purposes, as in, adding content and deleting, you could change the action depending on a URL param. That being said, obviously be extra careful not to delete content where you want to add it, and vice versa. 😅

As for the issue, I can't even get an empty DELETE request (without body) to work. 🥲

**Comment by leslee-glints** [comment]
> Just some clarity on method bodies. It's not common practice for methods other than PUT, POST and PATCH to have a 
> 
> There's no reason you can't have use a POST request for multiple purposes, as in, adding content and deleting, you could change the action depending on a URL param. That being said, obviously be extra careful not to delete content where you want to add it, and vice versa. 
> 
> As for the issue, I can't even get an empty DELETE request (without body) to work. 


### Working with next@13.4.4 by using POST Method facing this issue as well.

here are the error messages. 

**By useing request.body**
```javascript
export async function POST(request: Request) {
  const res = request.body;
}
```
Throws 
```
error StaticGenBailoutError: Page with `dynamic = "error"` couldn't be rendered statically because it used `request.body`.
```

**By useing request.json()**
```javascript
export async function POST(request: Request) {
  const res = request.json();
}
```
Throws 
```
error StaticGenBailoutError: Page with `dynamic = "error"` couldn't be rendered statically because it used `request.json`.
```


**By useing request.text()**
```javascript
export async function POST(request: Request) {
  const res = request.text();
}
```
Throws 
```
error StaticGenBailoutError: Page with `dynamic = "error"` couldn't be rendered statically because it used `request.text`.
```

### _Could anyone tell me how can I retrive BODY in from request ?_ 😿😿😿😿

**Comment by Amyx000** [comment]
> > Just some clarity on method bodies. It's not common practice for methods other than PUT, POST and PATCH to have a
> > There's no reason you can't have use a POST request for multiple purposes, as in, adding content and deleting, you could change the action depending on a URL param. That being said, obviously be extra careful not to delete content where you want to add it, and vice versa.
> > As for the issue, I can't even get an empty DELETE request (without body) to work.
> 
> ### Working with next@13.4.4 by using POST Method facing this issue as well.
> here are the error messages.
> 
> **By useing request.body**
> 
> ```js
> export async function POST(request: Request) {
>   const res = request.body;
> }
> ```
> 
> Throws
> 
> ```
> error StaticGenBailoutError: Page with `dynamic = "error"` couldn't be rendered statically because it used `request.body`.
> ```
> 
> **By useing request.json()**
> 
> ```js
> export async function POST(request: Request) {
>   const res = request.json();
> }
> ```
> 
> Throws
> 
> ```
> error StaticGenBailoutError: Page with `dynamic = "error"` couldn't be rendered statically because it used `request.json`.
> ```
> 
> **By useing request.text()**
> 
> ```js
> export async function POST(request: Request) {
>   const res = request.text();
> }
> ```
> 
> Throws
> 
> ```
> error StaticGenBailoutError: Page with `dynamic = "error"` couldn't be rendered statically because it used `request.text`.
> ```
> 
> ### _Could anyone tell me how can I retrive BODY in request ?_ 😿😿😿😿

use await `const body = await request.json()`

**Comment by leslee-glints** [comment]
Tried. Still doesn't work. 

![image](https://github.com/vercel/next.js/assets/130134817/83ac9d0e-74c3-4836-bc3b-d7cc9876b01b)

![image](https://github.com/vercel/next.js/assets/130134817/3920a874-77fc-4bc3-a2dd-8b1de78f4af1)

**Comment by RobertoGallego** [comment]
> Tried. Still doesn't work.
> 
> ![image](https://user-images.githubusercontent.com/130134817/242469159-83ac9d0e-74c3-4836-bc3b-d7cc9876b01b.png)
> 
> ![image](https://user-images.githubusercontent.com/130134817/242469251-3920a874-77fc-4bc3-a2dd-8b1de78f4af1.png)

I believe that this should work, maybe ur problem is coming from the way u call it, can u share with us your handle, and maybe try to change ur Response with NextResponse for example : 

```javascript
-    return new NextResponse("message", { status: 500 });


```

**Comment by mehtasag** [comment]
Hi guys ! I just found working solution to get ID when performing DELETE Request ......

`const deleteResponse = await fetch(`/api/yourRoute?id=${id}`, {
			method: "DELETE",

			headers: {
				"Content-Type": "application/json",
				"X-ID": id,
			},
		});`

Basically what we are passing ID in our own header... On other hand in NextJS 

> app/api/route we can access it by using 

```
const head = headers();

	console.log(request.headers);

	const head_token = head.get("x-id");

	console.log(head_token);

```
It works for me... Thanks

**Comment by leslee-glints** [comment]
> > Tried. Still doesn't work.
> > ![image](https://user-images.githubusercontent.com/130134817/242469159-83ac9d0e-74c3-4836-bc3b-d7cc9876b01b.png)
> > ![image](https://user-images.githubusercontent.com/130134817/242469251-3920a874-77fc-4bc3-a2dd-8b1de78f4af1.png)
> 
> I believe that this should work, maybe ur problem is coming from the way u call it, can u share with us your handle, and maybe try to change ur Response with NextResponse for example :
> 
> ```js
> return new NextResponse(data, { status: 200 });
> ```

Thank you for your useful advice. 

By changing to `return new NextResponse(data, { status: 200 });` it could work fine.

**Comment by BhargavEverestek** [comment]
![image](https://github.com/vercel/next.js/assets/43741543/c5b5e6aa-8cf9-4200-b2eb-55e9332cb106)

any solution??? 

> > Just some clarity on method bodies. It's not common practice for methods other than PUT, POST and PATCH to have a
> > There's no reason you can't have use a POST request for multiple purposes, as in, adding content and deleting, you could change the action depending on a URL param. That being said, obviously be extra careful not to delete content where you want to add it, and vice versa.
> > As for the issue, I can't even get an empty DELETE request (without body) to work.
> 
> ### Working with next@13.4.4 by using POST Method facing this issue as well.
> here are the error messages.
> 
> **By useing request.body**
> 
> ```js
> export async function POST(request: Request) {
>   const res = request.body;
> }
> ```
> 
> Throws
> 
> ```
> error StaticGenBailoutError: Page with `dynamic = "error"` couldn't be rendered statically because it used `request.body`.
> ```
> 
> **By useing request.json()**
> 
> ```js
> export async function POST(request: Request) {
>   const res = request.json();
> }
> ```
> 
> Throws
> 
> ```
> error StaticGenBailoutError: Page with `dynamic = "error"` couldn't be rendered statically because it used `request.json`.
> ```
> 
> **By useing request.text()**
> 
> ```js
> export async function POST(request: Request) {
>   const res = request.text();
> }
> ```
> 
> Throws
> 
> ```
> error StaticGenBailoutError: Page with `dynamic = "error"` couldn't be rendered statically because it used `request.text`.
> ```
> 
> ### _Could anyone tell me how can I retrive BODY in request ?_ 😿😿😿😿

**Comment by OlegKochiev** [comment]
+1:  work in production, don't work in localhost

**Comment by petrnovikov** [comment]
+1

**Comment by iv7dev** [comment]
Working with next@13.4.4 by using DELETE Method facing this issue as well.

> ### Verify canary release
> * [x]  I verified that the issue exists in the latest Next.js canary release
> 
> ### Provide environment information
> ```shell
> Operating System:
>       Platform: win32
>       Arch: x64
>       Version: Windows 10 Pro
>     Binaries:
>       Node: 18.12.1
>       npm: N/A
>       Yarn: N/A
>       pnpm: N/A
>     Relevant packages:
>       next: 13.3.1-canary.1
>       eslint-config-next: 13.2.4
>       react: 18.2.0
>       react-dom: 18.2.0
> ```
> 
> ### Which area(s) of Next.js are affected? (leave empty if unsure)
> _No response_
> 
> ### Link to the code that reproduces this issue
> N/A
> 
> ### To Reproduce
> ```ts
> import { NextResponse } from "next/server"
> const DATA_SOURCE_URL = "https://jsonplaceholder.typicode.com/todos"
> 
> type Todo = {
>     userId: number,
>     id: number,
>     title: string,
>     completed: boolean,
> }
> 
> export async function DELETE(request: Request) {
> 
>     const { id }: Partial<Todo> = await request.json() // error
> 
>     if (!id) return NextResponse.json({ "message": "Todo id required" })
> 
>     await fetch(`${DATA_SOURCE_URL}/${id}`, {
>         method: 'DELETE',
>     })
> 
>     return NextResponse.json({ "message": `Todo ${id} deleted` })
> }
> ```
> 
> ### Describe the Bug
> I only encounter this error on the DELETE route handler. My POST and PUT route handlers can `await request.json()` with no issues.
> 
> I run the same code in 13.2.4 and _do not_ receive the error.
> 
> I am sending valid and very simple JSON:
> 
> ```json
> {
>    "id": 3
> }
> ```
> 
> Nevertheless, I receive this error:
> 
> ```
> SyntaxError: Unexpected end of JSON input
>     at JSON.parse (<anonymous>)
>     at NextRequest.json (node:internal/deps/undici/undici:6160:23)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async DELETE (webpack-internal:///(sc_server)/./src/app/api/todos/route.ts:63:21)
> ```
> 
> ### Expected Behavior
> I expect `await request.json()` to work as it does in the POST and PUT route handlers.
> 
> ### Which browser are you using? (if relevant)
> _No response_
> 
> ### How are you deploying your application? (if relevant)
> _No response_
> 
> [NEXT-1194](https://linear.app/vercel/issue/NEXT-1194/after-upgrade-to-133-requestjson-in-delete-route-handler-fails)

+1 
Working with next@13.4.4 by using DELETE Method facing this issue as well.

**Comment by Amyx000** [comment]
> +1 Working with next@13.4.4 by using DELETE Method facing this issue as well.
> 
> > ### Verify canary release
> > 
> > * [x]   I verified that the issue exists in the latest Next.js canary release
> > 
> > ### Provide environment information
> > ```shell
> > Operating System:
> >       Platform: win32
> >       Arch: x64
> >       Version: Windows 10 Pro
> >     Binaries:
> >       Node: 18.12.1
> >       npm: N/A
> >       Yarn: N/A
> >       pnpm: N/A
> >     Relevant packages:
> >       next: 13.3.1-canary.1
> >       eslint-config-next: 13.2.4
> >       react: 18.2.0
> >       react-dom: 18.2.0
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ### Which area(s) of Next.js are affected? (leave empty if unsure)
> > _No response_
> > ### Link to the code that reproduces this issue
> > N/A
> > ### To Reproduce
> > ```ts
> > import { NextResponse } from "next/server"
> > const DATA_SOURCE_URL = "https://jsonplaceholder.typicode.com/todos"
> > 
> > type Todo = {
> >     userId: number,
> >     id: number,
> >     title: string,
> >     completed: boolean,
> > }
> > 
> > export async function DELETE(request: Request) {
> > 
> >     const { id }: Partial<Todo> = await request.json() // error
> > 
> >     if (!id) return NextResponse.json({ "message": "Todo id required" })
> > 
> >     await fetch(`${DATA_SOURCE_URL}/${id}`, {
> >         method: 'DELETE',
> >     })
> > 
> >     return NextResponse.json({ "message": `Todo ${id} deleted` })
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ### Describe the Bug
> > I only encounter this error on the DELETE route handler. My POST and PUT route handlers can `await request.json()` with no issues.
> > I run the same code in 13.2.4 and _do not_ receive the error.
> > I am sending valid and very simple JSON:
> > ```json
> > {
> >    "id": 3
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Nevertheless, I receive this error:
> > ```
> > SyntaxError: Unexpected end of JSON input
> >     at JSON.parse (<anonymous>)
> >     at NextRequest.json (node:internal/deps/undici/undici:6160:23)
> >     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> >     at async DELETE (webpack-internal:///(sc_server)/./src/app/api/todos/route.ts:63:21)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ### Expected Behavior
> > I expect `await request.json()` to work as it does in the POST and PUT route handlers.
> > ### Which browser are you using? (if relevant)
> > _No response_
> > ### How are you deploying your application? (if relevant)
> > _No response_
> > [NEXT-1194](https://linear.app/vercel/issue/NEXT-1194/after-upgrade-to-133-requestjson-in-delete-route-handler-fails)

You cant access request body for DELETE method, you can access the id from url.

Example route structure:  app/api/delete/[id]/route.ts

```
export async function DELETE(
  req: NextRequest,
  { params }: { params: { id: string } }
) {
  const { id } = params;
}
```

**Comment by bealecs** [comment]
+1

**Comment by elem15** [comment]
**Working with Next 13.4.4**

Only for POST, not for DELETE.

export async function POST(request: Request) {
  const body = await request.text()
  const { searchParams } = new URL(`http://api?${body}`)
  const obj = Object.fromEntries(searchParams.entries())
  return NextResponse.json(obj)
}

I hope it will be helpful

**Comment by james-william-r** [comment]
> **Working with Next 13.4.4**
> 
> Only for POST, PUT not for DELETE.
> 
> If you use JSON as Request Body
> 
> ```
> export async function POST(request: Request) {
>   const body = await request.text()
>   const obj = JSON.parse(body)
>   return NextResponse.json(obj)
> }
> ```
> 
> If you use 'x-www-form-urlencoded' as Request Body
> 
> ```
> export async function POST(request: Request) {
>   const body = await request.text()
>   const { searchParams } = new URL(`http://api?${body}`)
>   const obj = Object.fromEntries(searchParams.entries())
>   return NextResponse.json(obj)
> }
> ```
> 
> I hope it will be helpful

Unfortunately not, this question's specifically in relation to using DELETE in the route file. 😅

**Comment by mohamed-ali-agourram** [comment]
> **### The issue here is that the DELETE method typically expects the identifier of the resource to be deleted to be provided in the URL parameters, rather than in the request body. For example, you can include the id parameter in the URL like this:**

export async function DELETE(request: Request) {
    **const { searchParams } = new URL(request.url);
    const id = searchParams.get("id");**
    // Delete from the database using the extracted `id`
}

To perform the DELETE operation, you can make a request to the API endpoint with the id parameter specified in the URL:
   
export const delete_post = async (id: string) => {
    await fetch(`http://localhost:3000/api/posts?id=${id}`, {
        method: "DELETE",
        headers: {
            "Content-Type": "application/json",
        }
    });
}

**and this worked for me!**

**Comment by aniketbiprojit** [comment]
Encountered again. Fixed with editing `next.config.js`:

Doesn't work for static builds though which is sad.

<img width="379" alt="image" src="https://github.com/vercel/next.js/assets/24915107/7330991f-0b9b-4b68-a055-e1218f14f8ad">

**Comment by timneutkens** [comment]
@aniketbiprojit encountered what exactly? A fully static build can't have Route Handler that are not static, which is expected because there is no server.

**Comment by Schlumpf9** [comment]
Is it possible that this bug still exists using a custom server on 13.4.13? 
`await handle(req, res, parsedUrl);` fails with a 500 error "Internal Server Error" when executing a DELETE request. 

Even the handler function is not accessed doing a DELETE:
```
export default function apiHandler(req: NextApiRequest, res: NextApiResponse,) {
   logger.info('apiHandler');
   logger.info(`${req.method} ${req.url}`);
}
```

Using Next version 13.2.4 works just fine... I'm using the pages/api folder

**Comment by AlexanderFuse** [comment]
+1
13.4.12 + axios
Only working if you move DELETE method to serparate route, otherwise it uses GET instead of DELETE

**Comment by JohnFoster44** [comment]
This issue was resolved for me upgrading to 13.4.16
